File: 547-number-of-provinces/number-of-provinces.py
from collections import deque

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        ans = 0
        n = len(isConnected)
        q = deque()
        vis = [False for _ in range(n)]
        q.append(0)
        for i in range(n):
            if vis[i] == False:
                q.append(i)
                while q:
                    curr = q.popleft()
                    for j in range(n):
                        if isConnected[curr][j] == 1 and vis[j] == False:
                            q.append(j)
                            vis[j] = True
                ans += 1
        return ans

File: 547-number-of-provinces/number-of-provinces.java
class DisjointSet {

    int parent[];
    int size[];

    public DisjointSet(int n) {
        parent = new int[n + 1];
        size = new int[n + 1];

        for (int i=0; i<=n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    public int findParent(int node) {

        if (parent[node] == node) {
            return node;
        }

        return parent[node] = findParent(parent[node]);
    }

    public void unionBySize(int u, int v) {

        int parent_u = findParent(u);
        int parent_v = findParent(v);

        if (parent_u == parent_v) {
            return;
        }

        if (size[u] < size[v]) {
            parent[parent_u] = parent_v;
            size[parent_v] += size[parent_u];
        } else {
            parent[parent_v] = parent_u;
            size[parent_u] += size[parent_v];
        }
    }
}

class Solution {
    public int findCircleNum(int[][] isConnected) {

        int n = isConnected.length;
        DisjointSet ds = new DisjointSet(n);

        for (int i=0; i<n; i++) {
            for (int j=0; j<n; j++) {
                if (isConnected[i][j] == 1) {
                    ds.unionBySize(i, j);
                }
            }
        }

        int ans = 0;
        for (int i=0; i<n; i++) {
            if (ds.parent[i] == i) {
                ans++;
            }
        }

        return ans;
    }
}